distributed/service.py

```diff
@@ -0,0 +1,426 @@
+        ("output", 1): (service_result, "foo-service"),             # never actually runs
+        ("output", 2): (service_result, "foo-service"),             # never actually runs
+        ("output", 3): (service_result, "foo-service"),             # never actually runs
+        ("downstream", 1): (use, ("output", 1)),
+        ("downstream", 2): (use, ("output", 2)),
+        ("downstream", 3): (use, ("output", 3)),
+    }
+
+Or visually::
+
+    d1   d2   d3
+    |    |    |
+    o1   o2   o3
+      \  |   /
+    foo-service
+        / \
+      i1  i2
+
+If the single ``foo-service`` task is annotated as a service, it'll actually be computed
+by launching ``FooService`` instances on every worker, passing the inputs into them,
+letting them do whatever out-of-band operations they wish (including communicating with
+each other or external systems), and waiting for them to produce the output keys. Once
+each output key is produced (the `service_result` functions never actually run), any
+tasks downstream of it can run.
+
+---------------------------------------------------------------------------------------
+
+Dask handles startup/shutdown, resilience, peer discovery, leader election, RPC
+interfaces, and composition for Services:
+
+Startup/shutdown
+----------------
+
+* Once the *first* input key to a service task is ready (state ``memory``), service
+  instances are started on all workers in the cluster.
+* Once the last output key is produced by a service, all instances are told to stop.
+* Any instance can error the service task; all other instances will be told to stop.
+* Any instance can restart the service task; all other instances will be told to stop,
+  then re-launched from scratch.
+* If the service task is cancelled (because keys downstream from it are released, etc.),
+  all instances are told to stop.
+
+Resilience
+----------
+
+* If a worker that's running a service leaves, the entire service task is restarted on
+  all workers. TODO change once `peer_joined`/`peer_left` are added.
+* If keys downstream of a service need to be recomputed, the service will be rerun, just
+  like regular tasks are.
+* The service is informed at runtime which inputs to expect and which outputs to
+  produce, so partial recomputation (some outputs already in memory, some not) can be
+  handled correctly.
+
+Peer discovery and RPC
+----------------------
+
+* At startup, services are given a list of their peers as RPC handles.
+* The RPC handle allows a `Service` to call arbitrary methods on its peers on other
+  workers.
+* Loopback calls (using the RPC to call yourself) take a fastpath, bypassing
+  serialization and the network stack.
+
+Leader election
+---------------
+* At startup, one instance is designated as the "leader", making it easier to coordinate
+  initialization tasks that need to run exactly once.
+
+Composition
+-----------
+
+One Service's outputs can be another Service's inputs.
```

Comment:
I assume a join/merge would also be expressed as a service? I.e. two shuffle services and one combiner, correct?

This would remove our triangle fusion problem by transferring the responsibility of combining the correct shuffle outputs to produce a merged output. Does this combiner have all necessary information for this? What kind of metadata would we need to track and where to enable this, e.g. is this additional TaskState metadata, another out-of-band communication between the workers?

How would this look like for shuffling specifically?

Reply:
I wasn't thinking there'd be any special combiner service. Just two shuffle services, which feed into tasks combining their outputs.

```mermaid
flowchart BT
subgraph A
a1 & a2 & a3 --> shuffle-a --> sa1 & sa2 & sa3
end
subgraph B
b1 & b2 & b3 --> shuffle-b --> sb1 & sb2 & sb3
end
sa1 & sb1 --> m1
sa2 & sb2 --> m2
sa3 & sb3 --> m3
```

I still don't have an answer to triangle fusion. d84ce78aa4c528ac9c5918845648449f5ce3e1c7..ba08e615787a6d26b97fe3d8163d24faca858fb2 was an attempt to support triangle fusion natively in services, but it feels kind of contrived for the general Service interface.